### PR TITLE
#1320 - ScalarOf class created

### PR DESCRIPTION
**Pull request by marceloamadeu**
This PR is for #1320

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#1339](https://github.com/yegor256/cactoos/issues/1339) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/C63314D6Z/db6a210e-dac6-47db-b88f-b62b824aab1b); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/main/java/org/cactoos/scalar/ScalarOf.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * <p>-
+ * @param <T> Element type
+ * @since 0.4
+ * @todo #1320:30min Continue: Introduce a class ScalarOf that takes
+ *  various objects such as a Supplier, Callable, etc and implement
+ *  Scalar. Use it to replace in particular the casting to Scalar done
+ *  in ThreadsTest#failsDueToException, TimedTest#failsDueToException
+ *  and all the other places where it is needed. Add tests for each
+ *  of the constructors.
```
Comment (victornoel) [inline src/main/java/org/cactoos/scalar/ScalarOf.java:39]: @marceloamadeu can you at least use it in the two tests of the todo and rewrite the todo only for adding more constructors and to add tests?

File: src/main/java/org/cactoos/scalar/ScalarOf.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.cactoos.scalar;
```
Comment (victornoel) [inline src/main/java/org/cactoos/scalar/ScalarOf.java:24]: @marceloamadeu you forgot to remove the todo you are solving from the `package-info.java` file, you need to remove it for the job to be taken into account.

File: src/main/java/org/cactoos/scalar/ScalarOf.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+import org.cactoos.Scalar;
+
+/**
+ * ScalarOf.
+ *
+ * <p>-
```
Comment (victornoel) [inline src/main/java/org/cactoos/scalar/ScalarOf.java:31]: @marceloamadeu please remove this, this seems not needed

File: src/main/java/org/cactoos/scalar/ScalarOf.java (hunk 2 above)
Comment (victornoel) [reply src/main/java/org/cactoos/scalar/ScalarOf.java:24]: I'm not sure what's happening but you made two commit, one removing it, and one adding it...

**Comment by victornoel** [comment]
@marceloamadeu can you go on with this?

**Comment by victornoel** [comment]
@marceloamadeu ping?

**Comment by marceloamadeu** [comment]
@victornoel Please check the tests and todo.
Thank

**Comment by victornoel** [comment]
@marceloamadeu now you need to remove the todo you are solving (in the `package-info.java` file), if not the issue won't be closed automatically by 0pdd and you won't get paid :)

**Comment by marceloamadeu** [comment]
@victornoel Done... tks

File: src/main/java/org/cactoos/scalar/ScalarOf.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ *  objects such as a Supplier, Callable, etc and implement Scalar. Use it
+ *  to replace in all places where it is needed. Add tests for each of the
+ *  constructors.
+ */
+public final class ScalarOf<T> implements Scalar<T> {
+
+    /**
+     * The scalar.
+     */
+    private final Scalar<T> origin;
```
Comment (paulodamaso) [inline src/main/java/org/cactoos/scalar/ScalarOf.java:43]: @marceloamadeu This class seems a bit strange to me; shouldn't we be wrapping the `T` value, instead another scalar? @victornoel WDYT?

**Comment by paulodamaso** [comment]
@victornoel As reviewer, I want you opinon on this

**Comment by victornoel** [comment]
@paulodamaso the problem is that the initial todo was about introducing ScalarOf taking some of Java SDK corresponding classes (e.g. `Supplier`), so it can't just store the `T` but store something that will provide a `T`. For example the initial todo is replacing lambdas with `ScalarOf`, and the `ScalarOf` takes a lambda that throws an `Exception`.

A `Scalar<T>` field is not bad, even though it is a bit strange, but as long as it is hidden as a private field, anyway we can change it when needed.

What is more problematic is that the public constructor takes a `Scalar`, this is a strange situation I think, but maybe it's not so important...

An alternative to solve the initial todo being solved here would be to have a `ThrowingScalar` class that takes a `Scalar<Exception>` and that throws the Exception in its implementation. An then later we can add `SupplierScalar`, and so on...

**Comment by paulodamaso** [comment]
@victornoel I think that what we got here is enough. I'd like to see an implementation of `ScalarOf` which receives `T`, so we would be effectively wrapping the value, WDYT?

**Comment by victornoel** [comment]
@paulodamaso in that case it won't be usable as in the tests modified in this PR since they are wrapping lambda throwing exceptions. Or do you mean adding another constructor that takes T? In that case yes it will be useful for sure.

But it will increase ambiguity because what is the desired outcome of `new ScalarOf(() -> String)` :
- `Scara<Scalar<String>>`
- or `Scalar<String>`

**Comment by paulodamaso** [comment]
@victornoel I was thinking in adding another constructor, but I am concerned with this ambiguity problem. Lets keep it the way it is for now

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by victornoel** [comment]
@paulodamaso could you please tell 0crat out? it missed the event I think

**Comment by victornoel** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/1339#issuecomment-619370973))

@victornoel This is what I know about this job in C63314D6Z, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1339](https://github.com/yegor256/cactoos/issues/1339) is [in scope](http://www.zerocracy.com/policy.html#14) for 15days
  * The role is `REV`
  * The job is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel) for [15days](http://www.zerocracy.com/policy.html#8)
  * There is no monetary reward attached, it's a [free](http://www.zerocracy.com/policy.html#2) job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @marceloamadeu[/z](https://www.0crat.com/u/marceloamadeu): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%231339%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by paulodamaso** [comment]
@0crat out

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Code review was too long (15 days), architects (@paulodamaso) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by sereshqua** [comment]
@victornoel please make sure you start all your comments with the name of the user they are referred to, [see](https://github.com/yegor256/cactoos/pull/1339#discussion_r407047182)

**Comment by victornoel** [comment]
@sereshqua ha yes! thanks for the reminder :)

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Pull request merged**